### PR TITLE
Add k shortest paths

### LINKED ISSUE #1397: Feature request: K shortest paths
I would like to see "K shortest paths" functionality added to the core of igraph. This function would find the K shortest paths through a weighted graph using [Yen's algorithm](https://en.wikipedia.org/wiki/Yen%27s_algorithm).

The yenpathy package implements this, but as you can see in the issue tracker, there are several outstanding issues with the implementation that have not been addressed in several months (https://github.com/ecohealthalliance/yenpathy/issues). 

Thank you for your contributions to this community.

### PR DESCRIPTION
**Pull request by GroteGnoom**
Fixes #1397

Non-obvious things:
I use infinite weights to remove edges. This way I can use get_shortest_path_dijkstra directly.
I only give the path in edges. Vertices can be inferred from this afterwards.

### CODE REVIEW

**Comment by szhorvat** [comment]
I'm going to add some comments on this, but for now they are mostly self-reminders. No need to act on them yet.

File: src/paths/shortest_paths.c
Comment (szhorvat) [inline src/paths/shortest_paths.c:1384]: Do not change the item destructor permanently.

File: src/paths/shortest_paths.c
Comment (szhorvat) [inline src/paths/shortest_paths.c:1433]: `no_weights` to avoid confusion with 0 values.

**Comment by ntamas** [comment]
Resolved merge conflicts after the recent merge of `feat/typed-vector-lists` into `develop`.

The PR should be updated so that the new function uses `igraph_vector_int_list_t` to return the paths instead of an `igraph_vector_ptr_t`. I'll do this in a separate round once I'm done with checking the other PRs for merge conflicts.

**Comment by ntamas** [comment]
PR updated. One thing that I don't like about the current implementation is the fact that we are actively modifying the `weights` vector so it cannot be passed as a `const igraph_vector_t*` to the function, it must be non-const. The consequence is that I have to declare the parameter as `INOUT` in `functions.yaml`. Note that the function attempts to restore `weights` to its original state if it exist normally and not with an error code.

The alternatives would be to copy the `weights` vector before modifying it, and then working on the copy. In theory, the copying is O(m) so it does not affect the overall time complexity of the function, but it can still be problematic if the graph is large but the part actively modified during the traversal is small.

**Comment by ntamas** [comment]
Conflicts resolved with `develop`. I guess that this function should be called `igraph_get_k_shortest_paths()` now that we have removed `igraph_shortest_paths()` to `igraph_distances()`, right?

File: src/paths/shortest_paths.c
Comment (ntamas) [reply src/paths/shortest_paths.c]: Not relevant any more as we've switched to `igraph_vector_int_list_t`.

File: src/paths/shortest_paths.c
Comment (ntamas) [reply src/paths/shortest_paths.c:1441]: Renamed to `weights_missing` as `no_weights` can be mistaken for "number of weights".

**Comment by ntamas** [comment]
I think this is now ready to merge if the CI passes.

File: src/paths/shortest_paths.c
Comment (szhorvat) [inline src/paths/shortest_paths.c:1340]:
```suggestion
 * \brief k shortest paths between two vertices.
```

This function doesn't quite compute geodesics. Only the shortest path is a geodesic, not the 2nd shortest one.

File: src/paths/shortest_paths.c
Comment (szhorvat) [inline src/paths/shortest_paths.c:1352]:
```suggestion
 * \param from The ID of the vertex from which the paths are calculated.
```

File: src/paths/shortest_paths.c
Comment (szhorvat) [inline src/paths/shortest_paths.c:1354]:
```suggestion
 * \param to The ID of the vertex to which the paths are calculated.
```

File: src/paths/shortest_paths.c
Comment (szhorvat) [inline src/paths/shortest_paths.c:1379]:
```suggestion
 * Time complexity:  k |V| (|V| log|V| + |E|), where |V| is the number of vertices,
 *                  and |E| is the number of edges.
```

**Review by szhorvat — CHANGES_REQUESTED** [review]

File: tests/unit/igraph_get_k_shortest_paths.c
Comment (szhorvat) [inline tests/unit/igraph_get_k_shortest_paths.c:52]:
```suggestion
    igraph_small(&g_wiki, 6, IGRAPH_DIRECTED, 0,1, 0,2, 1,3, 2,1, 2,3, 2,4, 3,4, 3,5, 4,5, -1);
    igraph_small(&g_wiki_u, 6, IGRAPH_UNDIRECTED, 0,1, 0,2, 1,3, 2,1, 2,3, 2,4, 3,4, 3,5, 4,5, -1);
```

Just for my sanity.

Comment (szhorvat) [inline tests/unit/igraph_get_k_shortest_paths.c:52]:
```suggestion
    igraph_small(&g_0, 0, IGRAPH_UNDIRECTED, -1);
    igraph_small(&g_1, 1, IGRAPH_UNDIRECTED, -1);
    igraph_small(&g_2, 2, IGRAPH_UNDIRECTED, -1);
    igraph_small(&g_2c, 2, IGRAPH_UNDIRECTED, 0,1, -1);
    igraph_small(&g_wiki, 6, IGRAPH_DIRECTED, 0,1, 0,2, 1,3, 2,1, 2,3, 2,4, 3,4, 3,5, 4,5, -1);
    igraph_small(&g_wiki_u, 6, IGRAPH_UNDIRECTED, 0,1, 0,2, 1,3, 2,1, 2,3, 2,4, 3,4, 3,5, 4,5, -1);
```

Just for my sanity, there were too many numbers.

File: src/paths/shortest_paths.c
Comment (szhorvat) [inline src/paths/shortest_paths.c:1349]:
```suggestion
 *        object contains the edges along the <code>k</code>th shortest path between
 *         \p from and \p to, where \c k is the vector list index.
```

**Comment by szhorvat** [comment]
Are the paths returned in reverse? I am looking a the undirected Wikipedia example. The first result is `7, 4, 1`. But this is the path from `5` to `0`, not from `0` to `5`. It is reversed.

For reference, the graph is the following, with edge and vertex IDs marked (0-based):

image

**Comment by szhorvat** [comment]
Sorry, the function call was indeed from `5` to `0`, so it's fine.

**Comment by szhorvat** [comment]
Could we add some non-simple graphs to the tests?

image

image

File: CHANGELOG.md
Comment (szhorvat) [inline CHANGELOG.md:671]: This does not find `k` "alternative" shortestest paths, but simple _the `k` shortest paths_. They are not alternatives. They are the shortest, the 2nd shortest, the 3rd shortest, and so on.

File: src/paths/shortest_paths.c
Comment (szhorvat) [inline src/paths/shortest_paths.c:1365]:
```suggestion
 * \param mode The type of paths to be used for the
 *        calculation in directed graphs. Possible values:
 *        \clist
 *        \cli IGRAPH_OUT
 *          The outgoing paths of \p from are calculated.
 *        \cli IGRAPH_IN
 *          The incoming paths of \p from are calculated.
 *        \cli IGRAPH_ALL
 *          The directed graph is considered as an
 *          undirected one for the computation.
 *        \endclist
```

**Comment by szhorvat** [comment]
Are we sure we don't want to add an output argument to return the vertex paths? All other path functions have this, except `igraph_get_all_simple_paths`, which needs a complete refactoring anyway.

Note that the vertex paths can't be computed from the edge paths without also having at least the start vertex.

It would be a useful convenience to be able to return the vertex paths.

File: CHANGELOG.md
Comment (szhorvat) [reply CHANGELOG.md:671]:
```suggestion
 - `igraph_get_k_shortest_paths()` finds the k shortest paths between a source and a target vertex (#1763, thanks to @GroteGnoom)
```

Comment (szhorvat) [reply CHANGELOG.md:671]:
```suggestion
 - `igraph_get_k_shortest_paths()` finds the k shortest paths between a source and a target vertex (#1763, thanks to @GroteGnoom)
```

File: doc/structural.xxml
Comment (szhorvat) [inline doc/structural.xxml:27]: Can you move this to just before `get_all_simple_paths()`? It should not interrupt a block of closely related functions.

Comment (szhorvat) [inline doc/structural.xxml:27]: Can you move this to just before `get_all_simple_paths()`? It should not interrupt a block of closely related functions.

File: src/paths/shortest_paths.c
Comment (GroteGnoom) [reply src/paths/shortest_paths.c:1340]: Isn't geodesic just a synonym for shortest path? The second shortest path is not a geodesic in the same way as the second shortest path is not a shortest path.

Comment (GroteGnoom) [reply src/paths/shortest_paths.c:1340]: I do think it's fine without :)

**Review by szhorvat — APPROVED** [review]
@GroteGnoom Thanks for much for implementing this! This can be merged as far as I'm concerned.

I would still like to add an output argument for getting the vertex path, but I suggest doing that later, so this PR doesn't keep getting out of sync with develop.

**Pull request merged**
